src/pip/_internal/metadata/importlib/_compat.py

```diff
@@ -43,13 +46,21 @@ def get_info_location(d: importlib.metadata.Distribution) -> Optional[BasePath]:
     return getattr(d, "_path", None)
 
 
-def get_dist_name(dist: importlib.metadata.Distribution) -> str:
-    """Get the distribution's project name.
+def get_dist_canonical_name(dist: importlib.metadata.Distribution) -> NormalizedName:
+    """Get the distribution's normalized name.
 
     The ``name`` attribute is only available in Python 3.10 or later. We are
     targeting exactly that, but Mypy does not know this.
     """
+    # Try to get the name from the metadata directory name.
+    # This is much faster than reading metadata.
+    if info_location := get_info_location(dist):
+        stem, suffix = os.path.splitext(info_location.name)
+        if suffix in (".dist-info", ".egg-info"):
+            name = stem.split("-", 1)[0]
```

Comment:
This parsing logic can be extracted and shared by both name and version parsing code. It’d be easier to read too with early returns.